fl4health/clients/partial_weight_exchange_client.py

```diff
@@ -109,9 +113,15 @@ def set_parameters(self, parameters: NDArrays, config: Config) -> None:
                 unwound properly by the parameter exchanger
             config (Config): configuration if required to control parameter exchange.
         """
-        if not self.model_weights_initialized:
+        # Since PartialWeightExchangeClient may exchange different sets parameters
+        # across different training round, the full parameter exchanger is only used
+        # if a client is selected to participate in the first *fitting* round.
+        current_server_round = self.narrow_config_type(config, "current_server_round", int)
+        fl_fit_round = self.narrow_config_type(config, "fl_fit_round", bool)
+        if (not self.model_weights_initialized) and current_server_round == 1 and fl_fit_round:
             self.initialize_all_model_weights(parameters, config)
```

Comment:
Just placing a note here with the summary of our zoom conversation. Essentially, we decided that even for the BasicClient, we will only do FullParameterExchange on the first **fit** round of federated learning. That is, the server will fully sync all FL client models who participate in the first round of training. Thereafter, any models that join federated training for evaluation or subsequent training will only receive the weights transferred by their "correct" exchanger (partial or not). Get parameters will receive the boolean as to whether we're fitting or not from the fit or eval functions that call it and we'll know what round # we're on using the config

Reply:
> Essentially, we decided that even for the BasicClient, we will only do FullParameterExchange on the first fit round of federated learning. That is, the server will fully sync all FL client models who participate in the first round of training. 

Isn't this already the [default behaviour for BasicClient](https://github.com/VectorInstitute/FL4Health/blob/6472d0c477796edf1122943f6200e53f167ef6ab/fl4health/clients/basic_client.py#L151)?

> Thereafter, any models that join federated training for evaluation or subsequent training will only receive the weights transferred by their "correct" exchanger (partial or not).

If the correct exchanger is a FullParameterExchanger I assume we will still use it in subsequent rounds?

> Get parameters will receive the boolean as to whether we're fitting or not from the fit or eval functions that call it and we'll know what round # we're on using the config

All ParameterExchangers will follow this right? Have get_parameters take in the fitting bool? And Yuchong is going to update this to be true because its not currently the case?

Reply:
Let's talk about this a little more deeply together, but to answer your questions above
1) Sort of. We do use a FullParameterExchange whenever a client's weights are first initialized. This is problematic for partial weight exchange if a client does not participate in the first round of training.
2) Yes, we will always use the clients parameter exchange in subsequent rounds, as we do now. This includes when it's a FullParameterExchanger
3) We're only going to be changing the signature `get_parameters` function in the client. It won't change anything about how the ParameterExchangers function under the hood.